sample_factory/algorithms/appo/appo_utils.py

```diff
@@ -198,21 +198,22 @@ def cat(self, dict_of_tensor_arrays, macro_batch_size, use_pinned_memory, timing
                 tensor_batch = None
 
         if tensor_batch is None:
-            tensor_batch = copy_dict_structure(dict_of_tensor_arrays)
+            tensor_batch = copy_dict_structure(dict_of_arrays)
             log.info('Allocating new CPU tensor batch (could not get from the pool)')
 
-            for d1, cache_d, key, tensor_arr, _ in iter_dicts_recursively(dict_of_tensor_arrays, tensor_batch):
-                cache_d[key] = torch.from_numpy(np.concatenate(tensor_arr, axis=0))
+            for d1, cache_d, key, arr, _ in iter_dicts_recursively(dict_of_arrays, tensor_batch):
+                cache_d[key] = torch.from_numpy(arr)
                 if use_pinned_memory:
                     cache_d[key] = cache_d[key].pin_memory()
         else:
             with timing.add_time('batcher_mem'):
-                for d1, cache_d, key, tensor_arr, cache_t in iter_dicts_recursively(dict_of_tensor_arrays, tensor_batch):
-                    offset = 0
-                    for t in tensor_arr:
-                        first_dim = t.shape[0]
-                        cache_t[offset:offset + first_dim].copy_(torch.as_tensor(t))
-                        offset += first_dim
+                # this is slower than the older version where we copied trajectories to the cached tensor one-by-one
+                # this will only affect CPU-based envs with pixel observations, and won't matter after
```

Comment:
I'm assuming you've tested this, or have some data for which it will affect CPU-based envs only.

Reply:
Basically this is only slower if the learner's batcher is the bottleneck.
This only happens in rather extreme configurations (i.e. 100 CPU workers and 1 GPU learner).

Actually most of the slowdown that I measured was from slower conda environment rather than from this change.
Apparently newer versions of Numpy or MKL are quite a bit slower. The real slowdown is negligible for most envs.